CHANGELOG.md

```diff
@@ -51,6 +51,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * When creating a new workspace, the sparse patterns are now copied over from
   the current workspace.
 
+* Set config `snapshot.squash-consecutive-snapshots = true` to make consecutive
+  snapshots result in a single operation being recorded, effectively
+  automatically abandoning the intermediate snapshot operations, which helps
+  GC, especially if you run `jj log` in a loop for example.
```

Comment:
I don't think this is the best way to help GC intermediate snapshots. Instead, `jj util gc` (or `jj op <?>`) could have an option to squash operations flagged as snapshots?

Reply:
I mean, this is basically the same, no?.
Well, it's a config that does the squashing automatically (I guess I can add a command like you described, but have a config for it to happen automatically like here)

The benefit is that oplog is immediately less "spammy", and initially I thought it'd help with obslog, but I think that for that the jj-level gc has to run immediately for the replaced op - and maybe it's ok actually?. I should look into it, since three things I wanted from this are better oplog, better obslog and immediately gc-able intermediate snapshots

Reply:
My point is that nothing would be reclaimed until `jj util gc`, so it's safer to delay lossy operation until then.

I think `op log` could have an option to filter out repeated operations tagged as "snapshot". It's presentation problem, and can be solved without actually rewriting the history. `obslog` is yeah, different story as of now. Predecessors are tracked differently, and omitting operations doesn't help. All predecessor commits reachable from the operation heads won't be GC-ed.

Reply:
> safer to delay lossy operation until then

I see. Well, that's exactly why this is opt-in - personally, I'd want the hypothetical snapshot gc'ing (or abandoning, at least) command to be effectively run after every snapshot

> to filter out repeated operations

The presentation thing could actually happen by default, there's zero reasons against it (and then `op log` would have a flag to "show all snapshots") I think

> All predecessor commits reachable from the operation heads won't be GC-ed

Hm, wait, does that mean that in the current state of PR after I have 100 consecutive snapshots "squashed" all of the wc amends are still non-gc'able.

I need to research it more, it's the main motivation (and my main dumb gripe with jj, as I've been always saying) behind this, oplog presentation was a nice bonus

Reply:
> Hm, wait, does that mean that in the current state of PR after I have 100 consecutive snapshots "squashed" all of the wc amends are still non-gc'able.

Yes.

Reply:
Okay, so now I'm also doing the same thing to predecessors as to ops, which seems to work great for obslog, but idk if now those commits are gc'able

I can test that, I guess (at least manually verify)

But now I'm not sure how to extract it from (only being an) auto-action on snapshots when opted in to some op/util command - I think some form of predecessor gc should happen before that

Personally I'm double happy with how it looks, especially if the intermediate amends are indeed gc'able (aka garbage)

Reply:
It wouldn't be possible to extract predecessor thingy to `jj op` commands because they are commit metadata.

FWIW, I don't know why you have many intermediate snapshots. If `jj log` is running in background, it might be better to add an option to not commit the transaction.

Reply:
> many intermediate snapshots

Many is a strong word (the number 100 above was an over-the-top example ofc), I have a normal amount probably.

It just always bothered me that there are a bunch of redundant snapshots - it makes me avoid running readonly jj commands because they'd snapshot something super incomplete and super useless.

I know this is stupid but my brain just latches onto that, I've been whining about it on github and discord since forever and making ideas to fix it :)

This PR one such solution that I like a lot.

> to not commit the transaction

Actually this is something that Martin suggested in #2562 and hey would you look at that, the GC is implemented now :)

Adding the `--no-commit-transaction` flag should be very simple I guess.
And yes this is another solution that works, and one that definitely will be merged :)
Well the flag is a bit unwieldy, but I think there could be a config option to enable it for RO commands?.
~~if I'm not missing why this is obviously wrong~~

&nbsp;

Anyway I still like the idea in this PR a lot;
As I've said your point about it being a lossy operation that should happen with a command is covered by the fact that it's opt-in

It's ok if we decide not to have this in main, I'd just run my build then, like with signing 🙃

Reply:
> It's ok if we decide not to have this in main, I'd just run my build then, like with signing 🙃

Yes, I would prefer not to merge this. The `--no-commit-transaction` flag feels much more generic and thus better. We should still tag snapshot operations as we talked about.